### ISSUE TITLE
Save controller settings, pause while saving a screenshot, simplify PRG loading

### ISSUE THREAD

**Issue by dirkwhoffmann** [dirkwhoffmann/virtualc64#477]
User request: 

1) "Please add ability so save selected controllers on Port 1 and Port 2/ I'm selecting my gamepad on Porta 2 every time I start VirtualC64, because this setting is not saving itself."

2) "When I try to take a screenshot, game pauses for a moment and then starts running again. That pretty much ruins most of the screenshots, because dialog appears and I have to hit Enter and select a format maybe... Well, the moment is passed. I know, there are Quick Save Screenshot option, but still - every other emulators pauses games until screenshot is saved. And by the way, for some reason Alt+S does not work for me - in BASIC it's just produces this symbol (it attachement)."

3) "Can you please make running DSK, PRG and other files a little bit easier? Now I have to click on file in Finder, then click on first file (Enter does not work, it can only insert disk, for example), then press Enter again to execure "Run" command. Other emulator like Vise, while lacking in other areas (like saving screenshots or selecting other palletes), make this process so much simplier!"

**Comment by dirkwhoffmann** [comment]
1. I can save the settings in the user defaults. No problem.

2. When "Save screenshot ..:" is selected, I can pause emulation and continue when the save dialog disappears. Hitting Alt-s does not trigger a menu event, because all menu item shortcuts are prefixed with Cmd. Cmd-s is already reserved for "Save" and Alt-Cmd-s for "Save As ...". We are running out of key combinations here ... BTW, Alt-s produces the strange looking character, because the Alt-key is mapped to the Commodore key. 

3. The simplest way to quick-load a PRG is to enable "Skip user dialog when reading file" in the emulator preferences. If it is enabled, simply drag and drop a PRG file into the emulator window. It will automatically flash the file contents into memory and start the program.

**Comment by dirkwhoffmann** [comment]
I agree that taking screenshots should be made more flexible. In V3.3 the user will be able to choose freely if the emulator texture or the upscaled texture should be saved. The default image format will be "png" and not "tiff" any more. The choices are made in the emulator preferences, so they can be picked up by "Quicksave screenshot" as well. The new emulator preferences will look like this: 

<img width="615" alt="videosettings" src="https://user-images.githubusercontent.com/12561945/49526189-95bda880-f8af-11e8-9d29-5a82adf49642.png">

The video stuff will be moved to a separate video preferences dialog, because the current dialog is already crowded and it's likely that we need to add some new video options as well.

**Comment by dirkwhoffmann** [comment]
Regarding the shortcuts: "Quicksave screenshot" was indeed assigned to Alt-s in V3.1, a key combination that did not work. In V3.2, I changed it to Cmd-Alt-s. We cannot use Cmd-s for screenshots, because this combination is the default for saving the current document. I agree that "Hide sprites" does not need a shortcut. 

I could assign Cmd-Ctrl-s to "Save screenshot...", but is it really needed? Once the behaviour of "Quicksave screenshot" can be controlled in the emulator preferences and we fix the name thing, I guess that most users will prefer "Quicksave screenshot" to "Save screenshot..." anyway.

**Comment by michellekg** [comment]
> I agree that taking screenshots should be made more flexible. In V3.3 the user will be able to choose freely if the emulator texture or the upscaled texture should be saved.

Sorry for my ignoranceб but what is "texture"? Do you mean size of screenshot? Or something else?

> The default image format will be "png" and not "tiff" any more. The choices are made in the emulator preferences, so they can be picked up by "Quicksave screenshot" as well. The new emulator preferences will look like this:

That's very nice! Thanks. My only issue with this dialog is that snapshot preferences is grouped with screenshot ones. It can confuse somebody.

And by the way, is there plans to let users deside what folder screenshots (quick or normal ones) should be stored in?

> The video stuff will be moved to a separate video preferences dialog, because the current dialog is already crowded and it's likely that we need to add some new video options as well.

Totally agreed.

**Comment by michellekg** [comment]
> We cannot use Cmd-s for screenshots, because this combination is the default for saving the current document.

Yeah, your right. And power users can set this combination through System Preferences, so that's okay.

> I could assign Cmd-Ctrl-s to "Save screenshot...", but is it really needed? Once the behaviour of "Quicksave screenshot" can be controlled in the emulator preferences and we fix the name thing, I guess that most users will prefer "Quicksave screenshot" to "Save screenshot..." anyway.

Well, it would be nice to have at least some key combination (like, really anything), but you're probably right, Quick Screenshot with increment numbers in name etc is better anyway.

My only concert is that Save is Cmd+S and Take Snapshot is Cmd+T, so basically "saving state" is having two major key combinations right now (I know it's different, but you know what I mean) and screenshots having none. So maybe Cmd+T for Quick Screenshots and Cmd+Ctrl+S for Take Snapshots? Just a thought.

If it's by design (which I totally understand) so let it be. Although is better to decide which combination is easier to press to made access to Quick Screenshots easier too: Cmd+Alt+S or Cmd+Ctrl+S. For me it's the second one (with Ctrl), but maybe it's just me.

**Comment by dirkwhoffmann** [comment]
Feature is available in V3.3 alpha 1: 

http://www.dirkwhoffmann.de/virtualc64/VirtualC64_3.3_alpha1.zip

Please reopen if there is still an issue.

**Comment by michellekg** [comment]
Quoting from another topic:

> Regarding the name of a quick-saved screenshot, I have adapted the name scheme of Apple's grab tool:

Nice, although it works not exactly like Apple system. If multiple screenshots are made within one second, macOS names them as Screenshot 2018-12-07 at 12.10.56.png, Screenshot 2018-12-07 at 12.10.56 1.png, Screenshot 2018-12-07 at 12.10.56 2.png etc (so autoincrement after forst file), while VirtualC64 just replaces files. And sometimes you have to make many several screenshots in one seconds to choose a better shot.

Also controller settings is not saving after quitting program. Maybe I need to make other topic? TRhe same goes for quick opening of PRG, D64 and other files.

**Comment by michellekg** [comment]
Well, you're talking mostly about dragging flies to Emulator window, while I'm talking about opening them from Finder.

How about that - leave all that broad functionality to dragging files and open them from File menu, and assigning quick operations to opening files from browser?

So if I double click on PRG in Finder, these actions will be applied: 

a) flush PRG into memory
b) enter "RUN"
c) hit Enter

And with D64 it's like that:

a) open D64
b) flush first file on disk into memory

or:

a) open D64
b) type "LOAD "*",8,1:"
c) hit Enter
d) wait for it to load
e) type "RUN"
f) hit Enter

I don't know which is more accurate.

This way you're preserving current functionality (like opening disk 2), but simplifying opening 99% of games and apps.

**Comment by dirkwhoffmann** [comment]
I fear that we're running into a complexity problem here. Firstly, we have multiple ways to open a media file: 

- Menu -> Open ...
- Drag and drop
- Finder (double-clicking or dropping a file onto the dock icon)

Secondly, we have different file format that need to be handled differently. E.g., while flashing is the natural choice for PRGs and P00s, it almost never works for D64s or G64s. For those files, inserting as disk would be the natural choice. With the 7 supported file formats PRG, P00, T64, TAP, D64, G64, CRT, we end up with a choice matrix of 3 x 7 action items. Furthermore, I guess the user would not understand why adding a D64 via drag and drop does an auto-insert without auto-typing whereas opening the file from the finder would trigger auto-typing. 

My suggested solution is to add default actions for the 7 file formats in the emulator preferences. This bloats up the dialog a bit (what I don't like), but it makes clear what action is taken when a certain media file is opened.

<img width="645" alt="screenshot 2018-12-14 at 18 24 55" src="https://user-images.githubusercontent.com/12561945/50018195-bb912e80-ffce-11e8-9d69-819c062b78cd.png">

Option "Ask user" would bring up the mount dialogs as they come up right now, when option "Skip file browser when opening a media file" is unchecked. Hence, this option would no longer be needed.

**Comment by michellekg** [comment]
Sounds nice and not at all complicated.

Does "Insert into drive 8" means "Insert into drive 8, LOAD and RUN"?

**Comment by Alessandro1970** [comment]
_add default actions for the 7 file formats in the emulator preferences_
This is a good solution

**Comment by dirkwhoffmann** [comment]
I'm planning to make it a little more flexible: 

<img width="645" alt="screenshot 2018-12-15 at 14 47 03" src="https://user-images.githubusercontent.com/12561945/50043635-5f82e480-0078-11e9-9493-5ce1734eb34e.png">

Typing "RUN" after a file has been loaded from disk won't be possible though, because the emulator has no means to know when loading has finished.
